git/archive.go

```diff
@@ -0,0 +1,274 @@
+package git
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/flate"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+)
+
+type ArchiveOptions struct {
+	Verbose            bool
+	List               bool
+	WorktreeAttributes bool
+	ArchiveFormat      string
+	BasePrefix         string
+	OutputFile         string
+	CompressionLevel   int
+}
+
+type ArchiveFormat int
+
+const (
+	archiveUnknow = ArchiveFormat(iota)
+	archiveTar
+	archiveTarGzip
+	archiveZip
+)
+
+var supportedArchiveFormats = map[string]ArchiveFormat{
+	"tar":    archiveTar,
+	"tgz":    archiveTarGzip,
+	"tar.gz": archiveTarGzip,
+	"zip":    archiveZip,
+}
+
+func findOutputFileFormat(output string) ArchiveFormat {
+	// if the output is empty return the default value, a tarball.
+	if output == "" {
+		return archiveTar
+	}
+
+	for k, v := range supportedArchiveFormats {
+		if strings.HasSuffix(strings.ToLower(output), k) {
+			return v
+		}
+	}
+	// The archive format is not found, return unknow.
+	return archiveUnknow
+}
+
+func createTarArchive(c *Client, opts ArchiveOptions, tgz bool, sha Sha1, entries []*IndexEntry) error {
+	fileOutput := os.Stdout
+
+	// If the output file is not empty use it instead of stdout
+	if opts.OutputFile != "" {
+		if file, err := os.Create(opts.OutputFile); err == nil {
+			fileOutput = file
+			defer file.Close()
+		} else {
+			return err
+		}
+	}
+
+	mtime := time.Now()
+
+	// If the sha is a tree we set the file modification time to the current time
+	// otherwise we must use the commit time.
+	if sha.Type(c) != "tree" {
+		if t, err := CommitID(sha).GetDate(c); err == nil {
+			mtime = t
+		}
+	}
+
+	var tw *tar.Writer
+
+	//
+	if tgz {
+		gw := gzip.NewWriter(fileOutput)
+		defer gw.Close()
+
+		tw = tar.NewWriter(gw)
+		defer tw.Close()
+	} else {
+		tw = tar.NewWriter(fileOutput)
+		defer tw.Close()
+	}
+
+	// Write the pax header
+	hdr := &tar.Header{
+		Typeflag: tar.TypeXGlobalHeader,
+		Name:     "pax_global_header",
+		PAXRecords: map[string]string{
+			"comment": sha.String(),
+		},
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		o, err := c.GetObject(e.Sha1)
+		if err != nil {
+			return err
+		}
+		if obj, ok := o.(GitBlobObject); ok {
+			hdr := &tar.Header{}
+			hdr.Name = opts.BasePrefix + e.PathName.String()
+			hdr.Size = int64(obj.GetSize())
+			hdr.ModTime = mtime
+			hdr.Uid = 0
+			hdr.Gid = 0
+			hdr.Uname = "root"
+			hdr.Gname = "root"
```

Comment:
Are these permanent or just a stub? If a stub, there should probably be a comment explaining what real git does.

Reply:
The official client set the file owner and group name to root but this is probably not needed.